src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1435,29 +1435,39 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         wtp.paramTypes.foreach(instantiateSelected(_, tvarsToInstantiate))
         val constr = ctx.typerState.constraint
         def addImplicitArgs = {
-          def implicitArgError(msg: => String): Tree = {
-            ctx.error(msg, tree.pos.endPos)
-            EmptyTree
-          }
-          val args = (wtp.paramNames, wtp.paramTypes).zipped map { (pname, formal) =>
-            def where = d"parameter $pname of $methodStr"
-            inferImplicit(formal, EmptyTree, tree.pos.endPos) match {
-              case SearchSuccess(arg, _, _) =>
-                arg
-              case ambi: AmbiguousImplicits =>
-                implicitArgError(s"ambiguous implicits: ${ambi.explanation} of $where")
-              case failure: SearchFailure =>
-                implicitArgError(d"no implicit argument of type $formal found for $where" + failure.postscript)
-            }
+          def argsOrError(pnames: List[TermName], ptypes: List[Type], acc: mutable.ListBuffer[Tree])
+            : AnyRef /* List[Tree] | () => String */ = (pnames, ptypes) match {
+            case (pname :: pnames1, ptype :: ptypes1) =>
+              def where = d"parameter $pname of $methodStr"
+              inferImplicit(ptype, EmptyTree, tree.pos.endPos) match {
+                case SearchSuccess(arg, _, _) =>
+                  acc += arg
+                  argsOrError(pnames1, ptypes1, acc)
+                case ambi: AmbiguousImplicits =>
+                  () => s"ambiguous implicits: ${ambi.explanation} of $where"
```

Comment:
OK, intuitively I'd think that the cost of creating a closure would be higher than the cost of constructing a String